### DISCUSSION TITLE
How do I understand error messages?

### DISCUSSION THREAD

**Discussion by chokokatana** [trustwallet/wallet-core#3789] [Q&A] [answered]
When trying to build a `Cosmos.SigningInput.newBuilder()` and call the `AnySigner.sign(signingInput, CoinType, Cosmos.SigningOutput.parser())` I get the serialized empty string and the following messages in the android log:
```
error: Error_invalid_params
error_message: "Incorrect input parameter"
error_value: 22
```
Clearly something is wrong, but what precisely? Is the error value 22 pointing to something in my input structure?

**Comment by satoshiotomakan** [comment]
Hi @chokokatana, 22 refers to the `CommonError.Error_invalid_params` enum variant raw value.
Please share the `Proto.SigningInput` (except `private_key`) so I can help to find the root cause of the error

**Reply by chokokatana** [reply]
I'm using something like this:
```
val defaultFeeLimit = Cosmos.Amount.newBuilder().apply {
    amount = "5000"
    denom = "loki"
}.build()

val cosmosFee = Cosmos.Fee.newBuilder().apply {
    gas = 100000000L
    addAllAmounts(listOf(defaultFeeLimit))
}.build()

val transferMessage = Cosmos.Message.Transfer.newBuilder().apply {
    sourcePort = "transfer"
    sourceChannel = "channel-3"
    token = Cosmos.Amount.newBuilder().apply {
        amount = "1000"
        denom = "loki"
    }.build()
    sender = "odin1mjg8p9er2zkyteavzcqnctkuqaczeleuwlz4l8"
    receiver = "osmo1qxydza7ctzh9fn7sq5gcf0run8c78wh0jj47f4t5dwc5302r2dnqsp4hu6"
    timeoutTimestamp = 1712836904357727405
}.build()


val signingInput = Cosmos.SigningInput.newBuilder().apply {
    signingMode = Cosmos.SigningMode.Protobuf
    accountNumber = 53311L
    chainId = "odin-mainnet-freya"
    memo = "{\"wasm\":{\"contract\":\"osmo1qxydza7ctzh9fn7sq5gcf0run8c78wh0jj47f4t5dwc5302r2dnqsp4hu6\",\"msg\":{\"swap_and_action\":{\"user_swap\":{\"swap_exact_asset_in\":{\"swap_venue_name\":\"osmosis-poolmanager\",\"operations\":[{\"pool\":\"777\",\"denom_in\":\"ibc/C360EF34A86D334F625E4CBB7DA3223AEA97174B61F35BB3758081A8160F7D9B\",\"denom_out\":\"uosmo\"}]}},\"min_asset\":{\"native\":{\"denom\":\"uosmo\",\"amount\":\"55\"}},\"timeout_timestamp\":1712836904357747611,\"post_swap_action\":{\"transfer\":{\"to_address\":\"osmo1mjg8p9er2zkyteavzcqnctkuqaczeleuzyhx9l\"}},\"affiliates\":[]}}}}"
    sequence = 6L
    fee = cosmosFee
    privateKey = ByteString.copyFrom(odinKey.data())
    addAllMessages(listOf(Cosmos.Message.newBuilder().apply {
        transferTokensMessage = transferMessage
    }.build()))
}.build()

val output = AnySigner.sign(signingInput, CoinType.ODIN, Cosmos.SigningOutput.parser())
val outputString = output.serialized
```

**Reply by chokokatana** [reply]
Looking again through the [Cosmos.proto definition](https://github.com/trustwallet/wallet-core/blob/master/src/proto/Cosmos.proto#L66-L78) I notice I have not set the `timeout_height` field. The signing I'm trying to do uses data that comes from the [Skip API](https://api-docs.skip.money/docs/getting-started) and they simply provide `"timeout_height": {}` in their API. Maybe I have to set the `timeout_height` to some default value? If I try to set it to `null` it crashes.

On the other hand, if I remove the `addAllMessages(…)` I still get the `Error_invalid_params`, so I presume the error comes not from the messages but the _outer_ ones.

Also, the odin references may be confusing. These are from [this fork that includes Odin and other new tokens](https://github.com/GeoDB-Limited/wallet-core-odin/commit/aaa631a288207dd886ecb36f8aa8505400f6fa61).

Not sure if any of this helps or confuses more the issue. I tried grepping for `Error_invalid_params` and I find bits of code here and there but can't figure out where it is being actually generated.

**Reply by satoshiotomakan** [reply]
Hi @chokokatana, I'm taking a look right now, will update you shortly
Thanks for sharing!

**Reply by satoshiotomakan** [reply, marked as answer]
So you're right, `timeout_height` needs to be set.
If to remove all transaction messages, `Error_invalid_params` will be triggered because it's not allowed to generate an empty transaction.
Currently, we're working on error message improvements, so it will be easier to figure out what's the problem in the future.

Regarding the `timeout_height` values - recommendation is to set height, to rev. 1 and block current + 1000 (see api /blocks/latest)

**Reply by chokokatana** [reply]
Indeed, setting `timeout_height` allows the signing to be done. Thanks.

Regarding error messages, I captured the output through the log watching it in Android Studio. Is there any callback that can be set to capture these for me to redirect to Google Firebase? This would allow me to detect remote errors from production in the future.
